pkg/api/types.go

```diff
@@ -173,11 +173,121 @@ type VolumeSource struct {
-	// GitRepo represents a git repository at a particular revision.
-	GitRepo *GitRepo `json:"gitRepo"`
+
+	AWSElasticBlockStore *AWSElasticBlockStore `json:"elasticBlockStore"`
+
+	NFSMount *NFSMount `json:"nfsMount"`
+
+	PersistentVolumeClaimAttachment *PersistentVolumeClaimAttachment `json:"persistentVolumeClaim,omitempty"`
+}
+
+type PersistentVolume struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	//Spec defines a persistent volume owned by the cluster
+	Spec PersistentVolumeSpec `json:"spec,omitempty"`
+
+	// Status represents the current information about persistent volume.
+	Status PersistentVolumeStatus `json:"status,omitempty"`
+}
+
+type PersistentVolumeSpec struct {
+	Capacity ResourceList `json:"capacity,omitempty`
+	// AccessModeTypes are inferred from the Source
+	Source VolumeSource `json:"source,omitempty"`
+}
+
+type PersistentVolumeStatus struct {
+	Phase                          StoragePhase     `json:"phase,omitempty"`
+	PersistentVolumeClaimReference *ObjectReference `json:persistentVolumeClaimReference,omitempty`
+}
+
+type PersistentVolumeList struct {
+	TypeMeta `json:",inline"`
+	ListMeta `json:"metadata,omitempty"`
+	Items    []PersistentVolume `json:"items"`
+}
+
+// a PersistentVolumeClaim is a user's request for and claim to a persistent volume
+type PersistentVolumeClaim struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the volume
+	Spec PersistentVolumeClaimSpec `json:"spec,omitempty"`
+
+	// Status represents the current information about a persistent volume.
+	// this data may not be up to date.
+	Status PersistentVolumeClaimStatus `json:"status,omitempty"`
+}
+
+type PersistentVolumeClaimList struct {
+	TypeMeta `json:",inline"`
+	ListMeta `json:"metadata,omitempty"`
+	Items    []PersistentVolumeClaim `json:"items"`
+}
+
+// a PersistentVolumeClaimSpec describes the common attributes of storage devices
+// and allows a Source for provider-specific attributes
+type PersistentVolumeClaimSpec struct {
+	// Contains the types of access modes desired
+	AccessModes              AccessModeType    `json:"accessModes,omitempty"`
+	Resources                ResourceList      `json:"resources,omitempty"`
+	PersistentVolumeSelector map[string]string `json:"selector,omitempty"`
+}
+
+type PersistentVolumeClaimStatus struct {
+	Phase                     StoragePhase `json:"phase,omitempty"`
+	AccessModes               AccessModeType
+	Resources                 ResourceList
+	PersistentVolumeReference *ObjectReference
+}
+
+type ReadWriteOnce struct{}
+type ReadOnlyMany struct{}
+type ReadWriteMany struct{}
+
+type AccessModeType struct {
+	// Any of these access modes may be be specified.
+	// If none are specified, the default is ReadWriteOnce
+	ReadWriteOnce *ReadWriteOnce `json:"rwo,omitempty"`
+	ReadOnlyMany  *ReadOnlyMany  `json:"rox,omitempty"`
+	ReadWriteMany *ReadWriteMany `json:"rwx,omitempty"`
+}
+
+type StoragePhase string
```

Comment:
is Phase intended to go one-way only?  If so, how do we show status with regards to the many attach/mount unattach/unmount operations that may occur?